File: 04- April/16- Number of Ways to Form a Target String Given a Dictionary/16- Number of Ways to Form a Target String Given a Dictionary (Ahmed Gamal).js
// Author: Ahmed Gamal

// this is the iterative solution for this problem
// for the recursive solution and the explanation of the problem check my other solution in the same folder
// this solution is based on the idea of the recursive solution but it is implemented iteratively

/**
 * @param {string[]} words
 * @param {string} target
 * @return {number}
 */
var numWays = function(words, target) {
    // n is the length of each word
    // k is the length of the target string
    // MOD is the modulo value
    const n = words[0].length;
    const k = target.length;
    const MOD = 1000000007;

    // add function to add two numbers and return the result modulo MOD
    let add = (a, b) => {
        return (a + b) % MOD;
    };

    // freq[i][j] is the number of times the letter j appears in the ith position of all words
    
    // dp[i][j] is the number of ways to form the substring of target from index j to the end if we are at the ith row of the freq array
    // dp[i][j] = dp[i + 1][j] + j < k ? dp[i + 1][j + 1] * freq[i][target[j] - 'a'] : 0

    // from the transition function we can see that we only need the dp[i + 1] row to calculate the dp[i] row
    // and we only need to take the previous value of dp[i][j] and the value of dp[i][j + 1] to calculate the current value of dp[i][j]
    // so we can use a 1D array to store the dp values and update it in each iteration

    let freq = Array(n).fill(0).map(x => Array(26).fill(0));
    let dp = Array(k + 5).fill(0);

    // calculate the freq array
    words.forEach(word => {
        for(let i = 0; i < n; i++) {
            freq[i][word.charCodeAt(i) - 'a'.charCodeAt(0)]++;
        }
    });

    // initialize the base case
    dp[k] = 1;

    // iterate over the freq array from the last row to the first row
    for(let i = n - 1; ~i; --i) {
        // iterate over the dp array from the first column to the last column
        for(let j = 0; j <= k; j++) {
            // if j < k then we can take the current value of dp[j] and add to it the value of dp[j + 1] * freq[i][target[j] - 'a']
            if(j < k) {
                dp[j] = add(dp[j], dp[j + 1] * freq[i][target.charCodeAt(j) - 'a'.charCodeAt(0)]);
            }
        }
    }

    // return the value of dp[0]
    return dp[0];
};

File: 04- April/16- Number of Ways to Form a Target String Given a Dictionary/16- Number of Ways to Form a Target String Given a Dictionary (Ahmed Gamal).cpp
// Author: Ahmed Gamal

// for this problem, we can use DP to solve it
// since we not using any character in an index before the chosen character in a string, we can visualize the dictionary as an array of frequencies of each character in each index
// for example, if the dictionary is {"abc", "abd", "acd"}, then the array of frequencies will be:
// 3 0 0 0 0 0 .. 0 
// 0 2 1 0 0 0 .. 0
// 0 1 1 1 0 0 .. 0

// now, we can see that in each row of the array, we can choose the character we want to with frequency ways
// for example, if we want to choose the character 'a' in the first row, we can do it in 3 ways
// if we want to choose the character 'b' in the second row, we can do it in 2 ways
// if we want to choose the character 'c' in the third row, we can do it in 1 way
// and so on

// now, we can use DP to solve the problem
// dp[i][j] = number of ways to create the string target[j..] using the rows[i..] of the array of frequencies
// dp[i][j] = dp[i + 1][j] + (if j < target.size()) freq[i][target[j] - 'a'] * dp[i + 1][j + 1]


class Solution {
    // MOD = 1e9 + 7
    // n = number of rows in the array of frequencies
    // memo = memoization table
    // freq[i][j] = number of times the character 'a' + j appears in the row i of the array of frequencies
    const int MOD = int(1e9) + 7;
    int n;
    vector<vector<long long>> memo; 
    vector<vector<int>> freq;
    
    // add b to a, and take the modulo of the result by MOD
    void add(long long& a, long long b) {
        a = (a += b) % MOD;
    }
    
    // dp(i, j) = number of ways to create the string target[j..] using the rows[i..] of the array of frequencies
    long long dp(int idx, int curr, string& target) {
        // if we reached the end of the target string, then we can create it using the remaining rows of the array of frequencies
        if(idx >= n) {
            return curr == target.size();
        }

        // if the current state is already calculated, return the result
        long long& ret = memo[idx][curr];
        if(~ret) {
            return ret;
        }

        // dp(i, j) = dp(i + 1, j) + (if j < target.size()) freq[i][target[j] - 'a'] * dp(i + 1, j + 1)
        ret = dp(idx + 1, curr, target);
        if(curr < target.size()) {
            add(ret, 1ll * freq[idx][target[curr] - 'a'] * dp(idx + 1, curr + 1, target));
        }

        return ret;
    }
public:
    int numWays(vector<string>& words, string target) {
        // initialize your variables
        n = (int)words[0].size();
        memo.assign(n + 5, vector<long long>(target.size() + 5, -1));
        freq.assign(n + 5, vector<int>(26));

        // calculate the array of frequencies
        for(auto& word : words) {
            for(int c = 0; c < word.size(); c++) {
                freq[c][word[c] - 'a']++;
            }
        }

        // return the answer
        return dp(0, 0, target);
    }
};